OpenSim/Common/DataTable.h

```diff
@@ -0,0 +1,1142 @@
+      throw ZeroElements{"Input iterators produce zero elements."};
+    if(ndir == 0)
+      throw InvalidEntry{"Input argument 'ndir' cannot be zero."};
+
+    int row{0};
+    int col{0};
+    while(first != last) {
+      data.set(row, col, *first);
+      ++first;
+      if(dir == ROWWISE) {
+        ++col;
+        if(col == static_cast<int>(ndir)  && first != last) {
+          col = 0;
+          ++row;
+          data.resizeKeep(data.nrow() + 1, data.ncol());
+        }
+      } else {
+        ++row;
+        if(row == static_cast<int>(ndir) && first != last) {
+          row = 0;
+          ++col;
+          data.resizeKeep(data.nrow(), data.ncol() + 1);
+        }
+      }
+    }
+    if(!allow_missing) {
+      if((dir == ROWWISE && col != data.ncol()) ||
+         (dir == COLWISE && row != data.nrow()))
+        throw InvalidEntry{"Input iterator did not produce enough elements to "
+                           "fill the last row."};
+    }
+  }
+
+  /// Copy constructor. Perform a deep copy of the DataTable_.
+  DataTable_(const DataTable_& dt) :
+    data{dt.data}, metadata{dt.metadata}, col_ind{dt.col_ind} {
+      // Deep copy of metadata.
+      for(auto& elem : metadata)
+        elem.second = elem.second->clone();
+    }
+
+  /// Virtual copy constructor. Perform a deep copy of the DataTable_.
+  std::unique_ptr<AbstractDataTable> clone() const override {
+    return std::unique_ptr<AbstractDataTable>(new DataTable_{*this});
+  }
+
+  /// Copy assignment operator. Perform a deep copy of the DataTable_.
+  DataTable_& operator=(const DataTable_& dt) {
+    // Release metadata memory.
+    for(auto& elem : metadata)
+      delete elem.second;
+
+    data     = dt.data;
+    metadata = dt.metadata;
+    col_ind  = dt.col_ind;
+
+    // Deep copy of metadata.
+    for(auto& elem : metadata)
+      elem.second = elem.second->clone();
+
+    return *this;
+  }
+
+  /// Move constructor.
+  DataTable_(DataTable_&& dt) = default;
+
+  DataTable_& operator=(DataTable_&& dt) {
+    // Release metadata memory.
+    for(auto& elem : metadata)
+      delete elem.second;
+
+    data     = std::move(dt.data);
+    metadata = std::move(dt.metadata);
+    col_ind  = std::move(dt.col_ind);
+
+    return *this;
+  }
+
+  /// Destructor.
+  virtual ~DataTable_() {
+    // Release metadata memory.
+    for(auto& elem : metadata)
+      delete elem.second;
+  }
+
+  /// Get number of rows in the DataTable_.
+  size_t getNumRows() const {
+    return static_cast<size_t>(data.nrow()); 
+  }
+
+  /// Get number of columns in the DataTable_.
+  size_t getNumCols() const {
+    return static_cast<size_t>(data.ncol()); 
+  }
+
+  /// Get a row of the DataTable_ by index. Returned row is read-only. Use 
+  /// updRow to obtain a writable reference. See SimTK::RowVectorView_ for more
+  /// details.
+  ///
+  /// \throws SimTK::Exception::IndexOutOfRange If the row index specified
+  ///                                           is not in the DataTable_.
```

Comment:
This is probable not correct in Release builds. Indexing is expensive to check so is usually not checked unless requested (std::vector `operator[]` is unchecked while `at()` is checked). Another awkward problem with attempting to document `throw`s. It might make sense to check unconditionally when grabbing whole rows or columns but elementwise access can be very sensitive to per-element overhead.